### ISSUE TITLE
using oauth for token generation

### ISSUE THREAD

**Issue by cehbrecht** [bird-house/twitcher#22]

**Comment by cehbrecht** [comment]
use the ESGF OAuth service.

**Comment by tomLandry** [comment]
Label PAVICS please.

**Comment by cehbrecht** [comment]
Python libraries for OAuth.

Client: 
* https://github.com/requests/requests-oauthlib 

Provider:
* https://github.com/oauthlib/oauthlib

**Comment by cehbrecht** [comment]
An initial implementation using [pyramid-oauthlib](https://github.com/tilgovi/pyramid-oauthlib) is available on the `dev-oauth` branch.

The implementation is using a client-credentials workflow to get an access token:
https://oauthlib.readthedocs.io/en/latest/oauth2/grants/credentials.html 

Tokens are just simple strings with uuid:
https://oauthlib.readthedocs.io/en/latest/oauth2/tokens/tokens.html

The XMLRPC interface for service registration has been replaced by a Swagger/OpenAPI Rest interface protected by OAuth tokens using [Cornice](https://pythonhosted.org/cornice-swagger/index.html).

The resource access to OWS services (WPS, WMS, ...) is protected by OAuth tokens.

See the Jupyter notebook examples for usage.

See also [OAuth overview talk](https://www.slideshare.net/alvarosanchezmariscal/stateless-authentication-with-oauth-2-and-jwt-javazone-2015).

**Comment by cehbrecht** [comment]
@fmigneault Just be warned :) I won't merge to master before this is kind of stable and accepted. The adapter code needs to be tuned.

**Comment by fmigneault** [comment]
@cehbrecht 
👍 for Cornice, I like this approach.

What do you mean specifically about the access of OWS services with the token?
For instance, if I call `https://<host>/twitcher/ows/proxy/<service-name>`, I would need a token?
If so, will it be possible to have this token requirement disabled by config, as this is handled by Magpie (with it's own token/cookies) in PAVICS ?

**Comment by cehbrecht** [comment]
@fmigneault currently the WPS execute request can be blocked by an access token. But you can register a WPS as "public" and access is not restricted. The "access" is checked by twitcher ... it would just call an oauth `verify_request` method when configured.

**Comment by fmigneault** [comment]
@cehbrecht 
Can you point me to where this verification is accomplished? 
If the adapter interface is used along the way, maybe this won't cause any issue on Magpie side.
In fact, I think no actual Twitcher service gets created in our case since services are fetched from Magpie's db.

**Comment by cehbrecht** [comment]
@fmigneault currently the "access" check is done here:
https://github.com/bird-house/twitcher/blob/943b568253cf28874054005e59ba88ffc6b811a8/twitcher/owsproxy.py#L167

**Comment by fmigneault** [comment]
Looking at the process execution chain, to keep things working, I will need to add `public=True` here: 
https://github.com/Ouranosinc/Magpie/blob/63a7e2c39acfc85244e5f84ecb1622e9092a3fc1/magpie/adapter/magpieservice.py#L55-L57

I see however another issue, related to this call:
https://github.com/bird-house/twitcher/blob/943b568253cf28874054005e59ba88ffc6b811a8/twitcher/owsproxy.py#L169

Since Magpie defines other service and request types, the configs here are problematic.
https://github.com/bird-house/twitcher/blob/943b568253cf28874054005e59ba88ffc6b811a8/twitcher/owsrequest.py#L16-L26

One workaround could be to have these config variables defined in a separate file, which can be overridden in Dockerfile, similarly to how it's done with `celeryconfig.py`.
Another way would be to have a `ows_request_factory` defined in the adapter, which by default returns `OWSRequest`, but Magpie's adaptation would provide it's variant that always returns `true` when calling `OWSRequest.service_allowed()`.

**Comment by cehbrecht** [comment]
fixed in PR #86
